### DISCUSSION TITLE
Pipeline results in `pnpm: ENOTEMPTY: directory not empty` error in GitHub Action runs

### DISCUSSION THREAD

**Discussion by kevinpruett-prime** [vercel/turborepo#9067] [Help]
### Summary

Ever since updating to Turbo v2, we've seen our tasks result in issues related to the following error message:

```
ENOTEMPTY  ENOTEMPTY: directory not empty, rmdir '/home/runner/.local/share/pnpm/nodejs/20.17.0/lib/node_modules/npm/node_modules/<package>/node_modules'
```

Our `turbo.json` is as follows:

```json
{
  "$schema": "https://turbo.build/schema.json",
  "globalDependencies": [".env", ".env.local"],
  "tasks": {
    "build": {
      "dependsOn": ["^build"],
      "outputs": [
        "dist/**",
        "storybook-static/**",
        ".next/**",
        "!.next/cache/**"
      ]
    },
    "test": {},
    "lint": {},
    "dev": {
      "cache": false,
      "persistent": true
    },
    "clean": {
      "cache": false
    }
  }
}
```

Our root `package.json` is as follows:

```json
{
  "name": "<repo>",
  "devDependencies": {
    ...
    "turbo": "2.0.14"
  },
  "engines": {
    "node": "20.12.2"
  },
  "packageManager": "pnpm@9.9.0+sha512.60c18acd138bff695d339be6ad13f7e936eea6745660d4cc4a776d5247c540d0edee1a563695c183a66eb917ef88f2b4feb1fc25f32a7adcadc7aaf3438e99c1",
  "scripts": {
    "build": "turbo run build",
    "ci": "turbo run test lint",
    "clean": "turbo run clean && rm -rf node_modules",
    "deploy": "turbo run build",
    "dev": "turbo run dev --concurrency 11",
    "lint": "turbo run lint",
    "test": "turbo run test"
  }
}
```

When running `pnpm run ci` in GitHub Actions, for example, we receive the error highlighted above. **_Instead, when we run it explicitly as seen below, everything seems to work without incident; however I don't believe this to be a long-term solution as it seems to defeat the purpose of a tool like turbo:_**

```yaml
      - name: Run CI
        run: |
          pnpm run test --filter @<repo>/<packageA>
          pnpm run lint --filter @<repo>/<packageA>
          pnpm run test --filter=@<repo>/<packageB>
          pnpm run lint --filter=@<repo>/<packageB>
          pnpm run test --filter=@<repo>/<packageC>
          pnpm run lint --filter=@<repo>/<packageC>
          ...
```

### Additional information

FWIW, here is an example of what a `lint` script looks like in a given workspace

```json
  "scripts": {
    "lint": "eslint . --ext ts,tsx --report-unused-disable-directives --max-warnings 0"
  },
``` 

Similarly, the result of running `turbo run build` produced the same `ENOTEMPTY` issue, it is not isolated to the `test`, `lint`, or `build` jobs specifically, it appears to happen across the board.

**Comment by AlexErrant** [comment]
I have a similar issue with a [`tsc --build`](https://github.com/AlexErrant/Pentive/blob/0dd59c9a3b6a0a197e7b710d57b7d62f266cd59f/turbo.json#L4-L7) task. I'm on the latest PNPM (9.9.0), Turborepo (2.1.1), and Node LTS (20.17.0). Notably, running the task locally works.

The most common error message looks like [this run](https://github.com/AlexErrant/Pentive/actions/runs/10740453544/job/29788570810)

```
cache miss, executing 39bf96e15efb07cc
Fetching Node.js 20.17.0 ...
 ENOTEMPTY  ENOTEMPTY: directory not empty, rmdir '/home/runner/.local/share/pnpm/nodejs/20.17.0/lib/node_modules/npm/node_modules/text-table/example'

pnpm: ENOTEMPTY: directory not empty, rmdir '/home/runner/.local/share/pnpm/nodejs/20.17.0/lib/node_modules/npm/node_modules/text-table/example'
    at Object.rmdirSync (node:fs:1215:11)
    at _rmdirSync (node:internal/fs/rimraf:260:21)
    at rimrafSync (node:internal/fs/rimraf:193:7)
    at node:internal/fs/rimraf:[25](https://github.com/AlexErrant/Pentive/actions/runs/10740453544/job/29788570810#step:6:26)3:9
    at Array.forEach (<anonymous>)
    at _rmdirSync (node:internal/fs/rimraf:250:7)
    at rimrafSync (node:internal/fs/rimraf:193:7)
    at node:internal/fs/rimraf:253:9
    at Array.forEach (<anonymous>)
    at _rmdirSync (node:internal/fs/rimraf:250:7)
```

However, I did get a "[directory not empty, rename](https://github.com/AlexErrant/Pentive/actions/runs/10738683395/job/29785078475#step:6:35)" when I enabled [debug logging](https://docs.github.com/en/actions/monitoring-and-troubleshooting-workflows/troubleshooting-workflows/enabling-debug-logging).

I was able to replicate the issue [with Turborepo 2.0.0](https://github.com/AlexErrant/Pentive/commit/79222d72be3a92627f8e753f22cd22dec018a990). Trying with v1.13.4 [failed](https://github.com/AlexErrant/Pentive/actions/runs/10739604238/job/29785753171) because the turbo.json format changed. I couldn't find docs describing Turbo v1.x - [related](https://github.com/vercel/turborepo/discussions/9069) - and so I stopped there (because I was never on v1.x and have nothing to roll back to).

**Reply by AlexErrant** [reply]
Ulgh. Won't help OP, but my [turbo.json](https://github.com/AlexErrant/Pentive/blob/0dd59c9a3b6a0a197e7b710d57b7d62f266cd59f/turbo.json#L4-L7) used [transit nodes](https://turbo.build/repo/docs/core-concepts/package-and-task-graph#transit-nodes-as-entry-points):

```json
{
  "$schema": "https://turbo.build/schema.json",
  "tasks": {
    "tsc-build": {
      "dependsOn": ["transit"],
      "outputs": ["lib/**"]
    },
    "transit": {
      "dependsOn": ["^transit"]
    }
  }
}
```

which is more or less what's recommended by [the docs](https://turbo.build/repo/docs/guides/tools/typescript#linting-your-codebase):

```json

{
  "tasks": {
    "topo": {
      "dependsOn": ["^topo"]
    },
    "check-types": {
      "dependsOn": ["topo"]
    },
  },
}
```

I started deleting a ton of code in an attempt to minify the problem scope, and eventually discovered that the task was running *out*-of-topological-order:

```
• Packages in scope: api-ugc, hub, shared, shared-dom, shared-edge
• Running tsc-build in 5 packages
• Remote caching disabled
hub:tsc-build
cache miss, executing e4960e533c459059
Fetching Node.js [20](https://github.com/AlexErrant/Pentive/actions/runs/10744153624/job/29800451609#step:6:21).17.0 ...
 ENOTEMPTY  ENOTEMPTY: directory not empty, rmdir '/home/runner/.local/share/pnpm/nodejs/20.17.0/lib/node_modules/npm/node_modules/ssri'
```

https://github.com/AlexErrant/Pentive/actions/runs/10744153624/job/29800451609

`hub` has workspace dependencies! `shared` should be running _first_ since it has no workspace deps. Changing my `tsc-build` task to make it recursive [makes it work](https://github.com/AlexErrant/Pentive/commit/5c99111ecc46a1684ebebacbc7096371d147c380):

```json
   "tsc-build": {
      "dependsOn": ["transit", "^tsc-build"],
      "outputs": ["lib/**"]
    },
```

This, of course, [is _not_ recommended](https://turbo.build/repo/docs/crafting-your-repository/configuring-tasks#dependent-tasks-that-can-be-ran-in-parallel):

> This works...but could be faster!

This is [the branch](https://github.com/AlexErrant/Pentive/tree/debugging-turbo) where I minified the problem.

Strangely, the initial run linked in my parent comment correctly runs `shared` first. No idea why.

**Comment by kevinpruett-prime** [comment]
Thanks for the additional context here @AlexErrant ! I'm unclear whether something similar is happnening in my case as I don't really have topological order, each package can build independently, but this is helpful to chew on.

**Reply by AlexErrant** [reply]
I finally found my root cause. I stared really, _really_, _***really***_ hard at the lines before the error:

```
cache miss, executing 39bf96e15efb07cc
Fetching Node.js 20.17.0 ...
 ENOTEMPTY  ENOTEMPTY: directory not empty, rmdir '/home/runner/.local/share/pnpm/nodejs/20.17.0/lib/node_modules/npm/node_modules/text-table/example'
```

Why am I fetching _Node_ during a _Turbo run_? I already have `actions/setup-node@v4` setting up Node.

Turns out I had set [`use-node-version`](https://pnpm.io/npmrc#use-node-version):

> Specifies which exact Node.js version should be used for the project's runtime. pnpm will automatically install the specified version of Node.js...

So I deleted the config and everything worked as expected, including the transit nodes. Yay.
